sklearn/discriminant_analysis.py

```diff
@@ -806,14 +806,17 @@ def fit(self, X, y):
             Xgc = Xg - meang
             # Xgc = U * S * V.T
             _, S, Vt = np.linalg.svd(Xgc, full_matrices=False)
-            rank = np.sum(S > self.tol)
-            if rank < n_features:
-                warnings.warn("Variables are collinear")
             S2 = (S ** 2) / (len(Xg) - 1)
             S2 = ((1 - self.reg_param) * S2) + self.reg_param
+            cov_reg = np.dot(S2 * Vt.T, Vt)
+            det = linalg.det(cov_reg)
+            if det < self.tol:
```

Comment:
Traditionally, `tol` is compared to the singular values for determining the rank. Can we use `np.linalg.matrix_rank` here and pass in `tol`? (Internally `matrix_rank` computes the svd)

Reply:
matrix rank is going to be expansive also. I think we should just use the results of the SVD as explained in the above comment.

Reply:
In your comment here: https://github.com/scikit-learn/scikit-learn/issues/14997#issuecomment-787834014 

> I think that when the covariance matrix is not full rank we should probably raise a LinAlgError at fit time explicitly recommending the user to increase regularization.

From looking at the code, `S` does not depend on `reg_param`, which means increasing the regularization would still result in the error message.

Reply:
How about using `S2` in the same way as `S`?

Reply:
I experimented with the proposed change when using `S2` instead of `S` the solution works as intended but there are two tests that use random data that the changes couldn't pass:
* test_common.py/test_estimators[Quadratic_DiscriminantAnalysis-check_estimators_dtype]
* test_common.py/test_estimators[Quadratic_DiscriminantAnalysis-check_dtype_object]  


both of them fit the model with random data then check the existence of methods or attributes. This random data raises the errors. As the quick fixes, I could:
* Change the LinAlgError to LinAlgWarning which would still warn the user of the problem
* Mark the tests as fail
Do you have any further recommendations?

Reply:
As for a work around, we can update:

https://github.com/scikit-learn/scikit-learn/blob/26b6f60cd40f682570a80a02eb6484c69de88354/sklearn/utils/estimator_checks.py#L566

To set `reg_param`.